plugins/chat/assets/javascripts/discourse/services/chat-composer-warnings-tracker.js

```diff
@@ -0,0 +1,151 @@
+import Service, { inject as service } from "@ember/service";
+import { ajax } from "discourse/lib/ajax";
+import discourseDebounce from "discourse-common/lib/debounce";
+import { bind } from "discourse-common/utils/decorators";
+import { mentionRegex } from "pretty-text/mentions";
+import { cancel } from "@ember/runloop";
+import { tracked } from "@glimmer/tracking";
+
+const MENTION_RESULT = {
+  invalid: -1,
+  unreachable: 0,
+  over_members_limit: 1,
+};
+
+const MENTION_DEBOUNCE_MS = 1000;
+
+export default class ChatComposerWarningsTracker extends Service {
+  @service siteSettings;
+
+  // Track mention hints to display warnings
+  @tracked unreachableGroupMentions = null; // Array
+  @tracked overMembersLimitGroupMentions = null; // Array
+  @tracked tooManyMentions = false;
+  @tracked mentionsCount = 0;
+  @tracked mentionsTimer = null;
+
+  // Complimentary structure to avoid repeating mention checks.
+  _mentionWarningsSeen = null; // Hash
+
+  constructor() {
+    super(...arguments);
+
+    this.unreachableGroupMentions = [];
+    this.overMembersLimitGroupMentions = [];
+    this._mentionWarningsSeen = {};
+  }
+
+  willDestroy() {
+    cancel(this.mentionsTimer);
+  }
+
+  @bind
+  debouncedTrackMentions(message) {
+    this.message = message;
+
+    this.mentionsTimer = discourseDebounce(
+      this,
+      this._trackMentions,
+      MENTION_DEBOUNCE_MS
+    );
+  }
+
+  @bind
+  _trackMentions() {
+    if (!this.siteSettings.enable_mentions) {
+      return;
+    }
+
+    const mentions = this._extractMentions(this.message);
```

Comment:
another benefit of passing message is that you don't need to make it an instance value anymore but just a local